### ISSUE TITLE
struct and union declarations

### ISSUE THREAD

**Issue by nrc** [rust-lang-nursery/fmt-rfcs#30]
I'd like to discuss generics separately (#29). I'd like to include unit and tuple structs here.

**Comment by nrc** [comment]
Single-line tuple structs are already covered in the guide: https://github.com/rust-lang-nursery/fmt-rfcs/blob/master/guide/guide.md#tuples-and-tuple-structs

**Comment by ubsan** [comment]
I prefer:

```rust
struct Foo /* <- must be on one line */ { /* <- on the first line if it fits, otherwise pull it down */
  b: B, /* <- indented once, must end with a trailing comma */
  long_name:
    LongName, /* <- if required, pull the type name down to the next line and indent once */
} /* <- on it's own line */
```

**Comment by nrc** [comment]
And to clarify - a trailing comma on the last field?

**Comment by joshtriplett** [comment]
@nrc Yes.  That avoids having to add a comma when adding another field, or deleting a comma when deleting a field.

**Comment by bluss** [comment]
What about structs with trait bounds?

```rust
struct Peekable<I>
    where I: Iterator,
               I::Item: Debug, // This is only to illustrate more bounds
{
    peek: Option<I::Item>,
    iter: I,
}
```

- If the where clause is treated as a list, it too should always have a trailing comma; new additions go on a new line.

**Comment by Screwtapello** [comment]
I'm not a fan of visual identing, and it seems like the tone in other issues is against it. I'd rather have:

```rust
struct Peekable<I> where // 'where' on the same line, following the example of '{'
    I: Iterator,         // one block indent, just like struct fields or any other list item
    I::Item: Debug,      // trailing comma, of course
{ // visually ends the 'where' items and begins the struct items
    peek: Option<I::Item>,
    iter: I,
}

**Comment by bluss** [comment]
A hanging `where` works well for functions and methods too (they often have longer initial lines), so taken together I prefer it for all of them for consistency. A `where` at the end of a line is far removed from the actual where clauses. It's also been the prevailing style inside the rust project itself, what I can see.

**Comment by nrc** [comment]
I'd like to also discuss where clauses separately, since they equally affect functions, traits etc., I created #38 for that.

**Comment by nrc** [comment]
I think this is ready for FCP, summarised in https://github.com/rust-lang-nursery/fmt-rfcs/issues/30#issuecomment-254325732.

**Comment by johnthagen** [comment]
@nrc I would like to take a crack at this one.

**Comment by nrc** [comment]
assigning to myself for mentoring and so we can see @johnthagen is working on it

**Comment by nrc** [comment]
Closed by #53
